### ISSUE TITLE
[Bug] Token IDs not accepted by JSON grammar 

### ISSUE THREAD

**Issue by dtkettler** [mlc-ai/mlc-llm#2223]
## 🐛 Bug

Perhaps I'm jumping the gun on this because it looks like JSON grammar support is still under development.  But I was trying to use this feature and whenever I do I get errors like the following about certain token IDs not being accepted by the gammar state matcher:

```
Traceback (most recent call last):
  File "/home/david/anaconda3/envs/mlc-test/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/home/david/anaconda3/envs/mlc-test/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/david/anaconda3/envs/mlc-test/lib/python3.11/site-packages/mlc_llm/serve/engine_base.py", line 812, in _background_loop
    self._ffi["run_background_loop"]()
  File "tvm/_ffi/_cython/./packed_func.pxi", line 332, in tvm._ffi._cy3.core.PackedFuncBase.__call__
  File "tvm/_ffi/_cython/./packed_func.pxi", line 263, in tvm._ffi._cy3.core.FuncCall
  File "tvm/_ffi/_cython/./packed_func.pxi", line 252, in tvm._ffi._cy3.core.FuncCall3
  File "tvm/_ffi/_cython/./base.pxi", line 182, in tvm._ffi._cy3.core.CHECK_CALL
  File "/home/david/anaconda3/envs/mlc-test/lib/python3.11/site-packages/tvm/_ffi/base.py", line 481, in raise_last_ffi_error
    raise py_err
  File "/workspace/mlc-llm/cpp/serve/threaded_engine.cc", line 155, in mlc::llm::serve::ThreadedEngineImpl::RunBackgroundLoop()
  File "/workspace/mlc-llm/cpp/serve/engine.cc", line 275, in mlc::llm::serve::EngineImpl::Step()
  File "/workspace/mlc-llm/cpp/serve/engine_actions/batch_decode.cc", line 123, in mlc::llm::serve::BatchDecodeActionObj::Step(mlc::llm::serve::EngineState)
  File "/workspace/mlc-llm/cpp/serve/request_state.cc", line 57, in mlc::llm::serve::RequestModelStateNode::CommitToken(mlc::llm::serve::SampleResult)
tvm.error.InternalError: Traceback (most recent call last):
  3: mlc::llm::serve::ThreadedEngineImpl::RunBackgroundLoop()
        at /workspace/mlc-llm/cpp/serve/threaded_engine.cc:155
  2: mlc::llm::serve::EngineImpl::Step()
        at /workspace/mlc-llm/cpp/serve/engine.cc:275
  1: mlc::llm::serve::BatchDecodeActionObj::Step(mlc::llm::serve::EngineState)
        at /workspace/mlc-llm/cpp/serve/engine_actions/batch_decode.cc:123
  0: mlc::llm::serve::RequestModelStateNode::CommitToken(mlc::llm::serve::SampleResult)
        at /workspace/mlc-llm/cpp/serve/request_state.cc:57
  File "/workspace/mlc-llm/cpp/serve/request_state.cc", line 57
InternalError: Check failed: (accepted) is false: Token id 82246 is not accepted by the grammar state matcher.
```

## To Reproduce

I am using the python API with LLMEngine and I set response_format={"type": "json_object"} and stream=False in the call to engine.chat.completions.create() 

## Expected behavior

I'm expecting JSON-formatted output and not an error.

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): Vulkan
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Manjaro Linux
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...): Vega 64
 - How you installed MLC-LLM (`conda`, source): python3 -m pip install --pre -U -f https://mlc.ai/wheels mlc-llm-nightly mlc-ai-nightly
 - Python version (e.g. 3.10): 3.11

**Comment by tqchen** [comment]
thanks @dtkettler can you also let us know what is the model and prompt? having an example script would be helpful

**Comment by dtkettler** [comment]
Sure.  I have run into this with every prompt I have tried, but here's a very simple example adapted from the basic usage in the MLC-LLM documentation:

```
from mlc_llm import LLMEngine

# Create engine
model = "HF://mlc-ai/Llama-3-8B-Instruct-q4f16_1-MLC"
engine = LLMEngine(model)

# Run chat completion in OpenAI API.
response = engine.chat.completions.create(
    messages=[{"role": "user", "content": """
Produce a JSON object with a list of US states and the largest cities in those states
The list should have the following keys:
'state' - Name of the state
'cities' - Another list containing city names"""}],
    model=model,
    response_format={"type": "json_object"},
    stream=False
)

print(response)

engine.terminate()
```

**Comment by Ubospica** [comment]
Thanks for the report @dtkettler Currently there are several issues with the llama3 because it changes the tokenizer a lot. That will be fixed soon in these days

**Comment by FreakTheMighty** [comment]
I'm seeing a similar error when using `phi-2-q4f16_1-MLC`

```
InternalError: Check failed: (accepted) is false: Token id 0 is not accepted by the grammar state matcher
```

**Comment by tqchen** [comment]
llama3 should be fixed by latest version

**Comment by dtkettler** [comment]
Hi,

I tried it out and yes Llama 3 runs without error now.  However I still run into a pretty serious issue with it.  If I specify a schema in the response_format it just...returns the schema back to me.  Like, the output is literally just the schema I gave it, not an actual response.  If I just change the model to something else with no changes to my code I do not encounter this behavior so it's something specific to Llama 3.
